### ISSUE TITLE
Unable to start qube VMs if /var/log is empty

### ISSUE THREAD

**Issue by icequbes1** [QubesOS/qubes-issues#6090]
**Qubes OS version**
R4.0

**Affected component(s) or functionality**
qubesd, qubesdb-daemon, qubes-qmemman

**Brief summary**
No qubes will start after a reboot if /var/log directory is empty.

**To Reproduce**

In dom0 I did the following:

1. Remove all files inside of /var/log (I did this to free up space and remove the clutter of thousands of disposable VM log files)
2. reboot the computer
3. log back in
4. Observe errors when starting any qube

**Expected behavior**

Qubes should start normally when old log files do not exist.

**Actual behavior**

Absence of the /var/log/qubes directory and relevant permissions causes issues, as well as the absence of /var/log/qubes/qmemman.log.

**Screenshots**

**Additional context**

While an empty `/var/log` directory may be improbable, I would expect the OS to boot normally and not depend on the existence of log files in order to boot.

**Solutions you've tried**

I've gotten around the issue by adding to the `/usr/lib/qubes/fix-dir-perms.sh` script which runs at boot:

Inserting at the top of `fix-dir-perms.sh`:

```
mkdir -p /var/log/xen/console 2>/dev/null
mkdir -p /var/log/qubes 2>/dev/null
```

Appending to the bottom of `fix-dir-perms.sh`:

```
chgrp qubes /var/log/qubes
chmod 2770 /var/log/qubes
touch /var/log/qubes/qmemman.log
```

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by fepitre** [comment]
When something like this is happening, generally it means something bad happened. In many production systems, it can be a proof that you lost your log partition and this is absolutely not a good idea to force things booting before doing investigations or rescue operations like your did. If it's an issue with space or something there is proper solution like `logrotate`. I would rather add documentation about what Qubes logs can be safely deleted and those not. I've never seen some recommended procedure to `rm -rf /var/log/*` without ensuring critical services relies on some stamps into. @andrewdavidwong @marmarek

**Comment by icequbes1** [comment]
I agree you will never see `rm /var/log/*` as a recommended procedure! Only users that have thousands of disposable VM log files sitting around will do that and wipe everything :)

But even if all of /var/log is still there but certain Qubes log files removed, it seems odd to me for Qubes services to rely on log files already existing. Unless the idea is to consider inability to write log files at service start an uncoverable error?

**Comment by fepitre** [comment]
> But even if all of /var/log is still there but certain Qubes log files removed, it seems odd to me for Qubes services to rely on log files already existing. Unless the idea is to consider inability to write log files at service start an uncoverable error?

As I said, not specifically to Qubes services. In a more general way, some service assume that their log file is created before it starts. Even if it's inconvenient, some time it's a proof of a healthy log system, assuming the admin has properly setup the corresponding service and checked that everything is normal before starting it. It does matter because a service has no way for knowing if it's really suppose to recreate log file which potentially will have very huge size. Meaning in a failure case (lost of your log partition of such), you can literally crash and offline your production system.

> I agree you will never see `rm /var/log/*` as a recommended procedure! Only users that have thousands of disposable VM log files sitting around will do that and wipe everything :)

Yes we can add documentation about what can be safely removed.

**Comment by marmarek** [comment]
>  it seems odd to me for Qubes services to rely on log files already existing

They are not. They rely on specific _directories_ existing, with correct permissions. This standard practice, especially for services running as non-root user, as those cannot create directories themselves.

That said, it should be harmless enough to add relevant mkdir commands to `fix-dir-perms.sh` script by default.

**Comment by andrewdavidwong** [comment]
If I understand correctly: Qubes services follow a standard, general Linux practice of requiring appropriate log directories to exist. The user decided to manually delete these directories, resulting in Qubes services breaking. If this is correct, then:

1. This is not Qubes-specific. There is no sense in the Qubes OS Project trying to write documentation about Linux in general or about other projects when it is hard enough to write and maintain documentation for the original code that we produce.

2. This is an action that users shouldn't perform in the first place. There is a nearly infinite number of possible ways to break a Linux installation. There is no sense in trying to document them all.

Is the real problem that logs are not being cleaned up appropriately? Are they piling up and taking up an unreasonable amount of space? If so, then that should be the bug report, and the fix should be for the system to clean up after itself appropriately.

Alternatively, is it just that the user has a very small SSD and is trying to squeeze out every last GB by deleting files that turn out to be important? I've been there and sympathize, but that's out of scope for qubes-issues.

> Yes we can add documentation about what can be safely removed.

IMHO, users generally shouldn't be manually manipulating system files in dom0, and they should never *have* to in order for the system to function properly. Logs that are safe to remove should be cleaned up automatically with some sensible retention defaults. I won't go so far as to say that the documentation shouldn't be written, but there's a lot of higher-priority documentation that should be written first.

> That said, it should be harmless enough to add relevant mkdir commands to `fix-dir-perms.sh` script by default.

This seems very generous and more than sufficient to address this issue.

**Comment by icequbes1** [comment]
I agree that when the trigger is `rm -rf /var/log`, that it's a scenario that I would frown upon as a "why do users do these sorts of things?!!?" and is probably not worth the time to look into when there are more important issues to deal with.

I also did not submit this issue lightly; while the reason behind an empty `/var/log` can be questionable, the intent of this issue is to ask if Qubes would consider not being able to start qubes or core services due to the lack of `/var/log/qubes` an error that should be handled. I'd argue this issue is a very minor low-priority issue that simply addresses robustness and error handling to avoid the _unexpected_ consequences of removing log files that prevent the system from working properly.

If someone was to go messing around with `/etc/qubes`, then I'd consider that a case that is "out of scope" for Qubes.

Perhaps my view of `/var/log` is that it is not integral to a functional system and by nature a volatile write-only data store. If certain permissions and expectations are expected for a location like this, I would assume the service in question is responsible for configuring that location at startup.

Starting with an empty `/var/log` at boot, I see `/var/log/libvirt`, `/var/log/lightdm`, `/var/log/salt` all get created and the system boots without any issue. Case in point, `fix-dir-perms.sh` already sets up perms on `/var/log/xen` to add setgid / chgrp qubes.

On a more technical "root cause" of why the services didn't start, further debugging shows:

- Lack of `/var/log/qubes` causes qubesd, qubes-qmemman to not start (as those run as root)
   - I'd argue qubesd and qmemman are already going to the system journal anyway - /var/log/qubes/qubes.log and /var/log/qubes/qmemman.log are redundant.
   - qubes-qmemman starts _before_ the qubes-core service (the service ultimately executing `fix-dir-perms.sh`), so fixing permissions would need to happen earlier if that approach is taken.
- Lack of proper setgid/grp permissions of an existing `/var/log/qubes` cause qubesdb-daemon processes to fail (as those run as user)

I'm theory suggesting these approaches (I don't know enough about core concerns when it comes to logging for Qubes OS):

- Proposal 1: If services can't write to `/var/log/qubes` for whatever reason, do not consider it an error and keep executing to avoid the service failing to start.
- Proposal 2: Repair expected directory structure and permissions at boot in the appropriate location (whether fix-dir-perms.sh or somewhere else).
- Optional addition to either: Remove logging to `/var/log/qubes` for services that already get stored in the journal (such as qubesd and qubes-qmemman)

I am sort of questioning the necessity of documentation that states "don't remove these files". Where would that go and would users honestly even read it (before they've wiped the directory and are now searching docs for why qubes do not start)?

**Comment by andrewdavidwong** [comment]
Thanks for the clarifications, @icequbes1. I better understand where you're coming from.

> the intent of this issue is to ask if Qubes would consider not being able to start qubes or core services due to the lack of `/var/log/qubes` an error that should be handled. I'd argue this issue is a very minor low-priority issue that simply addresses robustness and error handling to avoid the _unexpected_ consequences of removing log files that prevent the system from working properly.

> Perhaps my view of `/var/log` is that it is not integral to a functional system and by nature a volatile write-only data store. If certain permissions and expectations are expected for a location like this, I would assume the service in question is responsible for configuring that location at startup.

It seems to me that these are the key things for the devs to decide whether they consider anything here a bug or not.

> I am sort of questioning the necessity of documentation that states "don't remove these files". Where would that go and would users honestly even read it (before they've wiped the directory and are now searching docs for why qubes do not start)? 

Agreed.
